torch/lib/THD/base/data_channels/DataChannelGloo.hpp

```diff
@@ -0,0 +1,304 @@
+  }
+
+  static GlooCache::value_type create(
+    const DataChannel::Group& group, DataChannelGloo:store_type& store
+    std::size_t input_bytes, std::size_t count, rank_type src_rank
+  ) {
+    auto& cache = GlooCache::get();
+    auto context = cache.createContext(group, store);
+    auto input_buffer = cache.createBuffer(input_bytes);
+    return std::make_tuple(
+      context,
+      std::make_shared<::gloo::BroadcastOneToAll<T>>(
+        *context, {reinterpret_cast<T*>(input_buffer->data()}, count, src_rank
+      ),
+      input_buffer,
+      input_buffer // we get the result in same buffer
+    );
+  }
+};
+
+template<typename T> // unused
+struct algorithm_spec<DataOperation::BARRIER, T> {
+  static GlooCache::key_type key(THDGroup group_id) {
+    return std::make_tuple(DataOperation::BARRIER, group_id, 0, 0, 0);
+  }
+
+  static GlooCache::value_type create(
+    const DataChannel::Group& group,
+    DataChannelGloo:store_type& store
+  ) {
+    auto context = GlooCache::get().createContext(group, store);
+    return std::make_tuple(
+      context,
+      std::make_shared<::gloo::BarrierAllToAll>(*context),
+      nullptr,
+      nullptr
+    );
+  }
+};
+
+
+struct GlooCache {
+  using key_type = std::tuple<
+    DataOperation, THDGroup, std::size_t, std::size_t, THDReduceOp
+  >;
+  using value_type = std::tuple<
+    std::shared_ptr<algorithm_type>, // algorithm
+    std::shared_ptr<buffer_type>,    // input buffer (nullptr if not used)
+    std::shared_ptr<buffer_type>     // output buffer (nullptr if not used)
+  >;
+
+  GlooCache(GlooCache const&)       = delete;
+  void operator=(GlooCache const&)  = delete;
+
+  // singleton instance
+  static GlooCache& get() {
+    static GlooCache instance;
+    return instance;
+  }
+
+  // TODO: enforce setters to be the first thing called (maybe some kind of factory)
+  void setRank(rank_type rank) {
+    _rank = rank;
+  }
+
+  void setDevice(std::shared_ptr<::gloo::transport::Device> device) {
+    _device = device;
+  }
+
+  std::make_shared<context_type> createContext(
+    const DataChannel::Group& group,
+    DataChannelGloo::store_type& store
+  ) {
+    auto context = std::make_shared<context_type>(group_rank, group.size());
+    context->connectFullMesh(store, _device);
+    return context;
+  }
+
+  std::shared_ptr<buffer_type> createBuffer(std::size_t bytes) {
+    return std::shared_ptr<buffer_type>(new char[bytes],
+                                        std::default_delete<buffer_type>());
+  }
+
+  template<DataOperation D, typename T, typename... Args>
+  value_type getAlgorithm(THDGroup group_id, const DataChannel::Group& group,
+                          DataChannelGloo:store_type& store, Args&&... args) {
+    auto key = algorithm_spec<D, T>::key(group_id, std::forward<Args>(args)...);
+    if (_algorithms.find(key) == _algorithms.end()) {
+      _algorithms[key] = algorithm_spec<D, T>::create(group, store, std::forward<Args>(args)...);
```

Comment:
I don't think so. They're moved when you forward them (see [SO thread](http://stackoverflow.com/questions/20195908/perfect-forwarding-with-multiple-passes-over-input-arguments)).